File: zig_compiler/runtime/io_ops_format.c
#include "basic_runtime.h"
#include "string_descriptor.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <stdbool.h>
#include <ctype.h>

// Internal helper: Reverses a string in place
static void reverse_str(char* s) {
    int i, j;
    for (i = 0, j = strlen(s) - 1; i < j; i++, j--) {
        char c = s[i]; s[i] = s[j]; s[j] = c;
    }
}

// Parse and format a numeric value according to a mask pattern
static void format_numeric(char* output, size_t output_size, const char* mask, const char* value_str) {
    // Try to parse value as a number
    char* endptr;
    double value = strtod(value_str, &endptr);
    bool is_numeric = (*endptr == '\0' || isspace(*endptr));
    
    if (!is_numeric) {
        // Not a number - just copy the value
        snprintf(output, output_size, "%s", value_str);
        return;
    }
    
    bool is_neg = (value < 0);
    double abs_val = fabs(value);

    // Analyze mask features
    bool has_comma = (strchr(mask, ',') != NULL);
    bool has_plus = (mask[0] == '+');
    bool has_minus_suffix = (mask[strlen(mask) - 1] == '-');
    bool has_exp = (strstr(mask, "^^^^") != NULL);
    bool has_dollar = (strstr(mask, "$$") != NULL);
    bool has_asterisk = (strstr(mask, "**") != NULL);

    // Determine precision
    int precision = 0;
    const char* dot = strchr(mask, '.');
    if (dot) {
        for (const char* p = dot + 1; *p == '#' || *p == '^'; p++) precision++;
    }

    // Core conversion
    char work[128];
    if (has_exp) {
        snprintf(work, sizeof(work), "%.*E", precision, value);
    } else {
        snprintf(work, sizeof(work), "%.*f", precision, abs_val);
    }

    // Manual comma insertion (integer part only)
    if (has_comma && !has_exp) {
        char *w_dot = strchr(work, '.');
        int int_len = w_dot ? (int)(w_dot - work) : (int)strlen(work);
        
        int src = int_len - 1;
        int dst = 0;
        int count = 0;
        char temp[128] = {0};

        while (src >= 0) {
            if (count > 0 && count % 3 == 0) temp[dst++] = ',';
            temp[dst++] = work[src--];
            count++;
        }
        reverse_str(temp);
        if (w_dot) strcat(temp, w_dot); // Append decimal part
        strcpy(work, temp);
    }

    // Decorations (Sign, $, *)
    char decorated[256] = {0};
    char prefix[10] = {0};
    
    if (has_plus) {
        strcat(prefix, is_neg ? "-" : "+");
    } else if (is_neg && !has_minus_suffix) {
        strcat(prefix, "-");
    }

    if (has_dollar) strcat(prefix, "$");
    
    snprintf(decorated, sizeof(decorated), "%s%s%s", prefix, work, (is_neg && has_minus_suffix) ? "-" : "");

    // Padding logic
    int mask_len = strlen(mask);
    int actual_len = strlen(decorated);
    int pad = mask_len - actual_len;

    if (pad < 0) {
        // Overflow - prefix with %
        snprintf(output, output_size, "%%%s", decorated);
    } else {
        // Pad with spaces or asterisks
        char padded[256] = {0};
        memset(padded, has_asterisk ? '*' : ' ', pad);
        strcat(padded, decorated);
        snprintf(output, output_size, "%s", padded);
    }
}

// Extract a format pattern starting at position p
// Returns the length of the pattern (0 if not a pattern)
static int extract_pattern(const char* p, char* pattern, size_t pattern_size) {
    const char* start = p;
    
    // Check for leading +, $, or *
    if (*p == '+' || (*p == '$' && *(p+1) == '$') || (*p == '*' && *(p+1) == '*')) {
        if (*p == '+') {
            pattern[0] = '+'; pattern[1] = '\0';
            p++;
        } else {
            pattern[0] = p[0]; pattern[1] = p[1]; pattern[2] = '\0';
            p += 2;
        }
    } else {
        pattern[0] = '\0';
    }
    
    // Collect #, comma, and decimal point
    size_t len = strlen(pattern);
    while (*p == '#' || *p == ',' || *p == '.') {
        if (len < pattern_size - 1) {
            pattern[len++] = *p;
            pattern[len] = '\0';
        }
        p++;
    }
    
    // Check for ^^^^
    if (strncmp(p, "^^^^", 4) == 0) {
        if (len + 4 < pattern_size) {
            strcat(pattern, "^^^^");
            len += 4;
        }
        p += 4;
    }
    
    // Check for trailing -
    if (*p == '-') {
        if (len < pattern_size - 1) {
            pattern[len++] = '-';
            pattern[len] = '\0';
        }
        p++;
    }
    
    // Return length only if we found at least one # or @
    if (strchr(pattern, '#') || strchr(start, '@')) {
        return (int)(p - start);
    }
    
    return 0;
}

// PRINT USING implementation with full numeric formatting support
void basic_print_using(StringDescriptor* format, int64_t count, StringDescriptor** args) {
    if (!format) return;

    // Extract format string UTF-8 immediately
    const char* fmt = string_to_utf8(format);
    if (!fmt) return;
    
    // Make a copy of format string since descriptors may be released
    char* fmt_copy = strdup(fmt);
    if (!fmt_copy) {
        basic_error_msg("Out of memory in basic_print_using");
        return;
    }

    // Collect argument UTF-8 strings immediately (before any descriptors are released)
    char** arg_strings = NULL;
    if (count > 0 && args) {
        arg_strings = (char**)calloc((size_t)count, sizeof(char*));
        if (!arg_strings) {
            free(fmt_copy);
            basic_error_msg("Out of memory in basic_print_using");
            return;
        }

        for (int64_t i = 0; i < count; i++) {
            if (args[i]) {
                const char* s = string_to_utf8(args[i]);
                if (s) {
                    arg_strings[i] = strdup(s);
                }
            }
        }
    }

    // Process the format string with collected arguments
    int64_t argIndex = 0;

    for (const char* p = fmt_copy; *p; ) {
        // Check for @ string placeholder
        if (*p == '@') {
            // String substitution - no special formatting
            if (argIndex < count && arg_strings && arg_strings[argIndex]) {
                printf("%s", arg_strings[argIndex]);
            }
            argIndex++;
            p++;
        }
        // Check for numeric format pattern
        else if (*p == '#' || *p == '+' || 
                 (*p == '$' && *(p+1) == '$') || 
                 (*p == '*' && *(p+1) == '*')) {
            char pattern[128];
            int pattern_len = extract_pattern(p, pattern, sizeof(pattern));
            
            if (pattern_len > 0) {
                // Apply numeric formatting
                if (argIndex < count && arg_strings && arg_strings[argIndex]) {
                    char formatted[256];
                    format_numeric(formatted, sizeof(formatted), pattern, arg_strings[argIndex]);
                    printf("%s", formatted);
                }
                argIndex++;
                p += pattern_len;
            } else {
                putchar(*p++);
            }
        } else {
            putchar(*p++);
        }
    }

    fflush(stdout);

    // Clean up
    if (arg_strings) {
        for (int64_t i = 0; i < count; i++) {
            if (arg_strings[i]) free(arg_strings[i]);
        }
        free(arg_strings);
    }
    free(fmt_copy);
}

File: zig_compiler/runtime/io_ops_format.zig
//
// io_ops_format.zig
// FasterBASIC Runtime — PRINT USING Implementation
//
// Full numeric formatting support: masks with #, commas, $$, **,
// +, trailing -, and ^^^^ (scientific notation).
//

const std = @import("std");
const c = std.c;

// =========================================================================
// Extern declarations
// =========================================================================

extern fn basic_error_msg(msg: [*:0]const u8) void;
extern fn string_to_utf8(str: ?*anyopaque) [*:0]const u8;

// C library
extern fn snprintf(buf: [*]u8, size: usize, fmt: [*:0]const u8, ...) c_int;
extern fn strtod(nptr: [*:0]const u8, endptr: ?*[*:0]const u8) f64;
extern fn strdup(s: [*:0]const u8) ?[*:0]u8;
extern fn printf(fmt: [*:0]const u8, ...) c_int;
extern fn putchar(ch: c_int) c_int;
extern fn fflush(stream: ?*anyopaque) c_int;
extern fn strchr(s: [*:0]const u8, ch: c_int) ?[*:0]const u8;
extern fn strstr(haystack: [*:0]const u8, needle: [*:0]const u8) ?[*:0]const u8;
extern fn strncmp(s1: [*:0]const u8, s2: [*:0]const u8, n: usize) c_int;
extern fn strlen(s: [*:0]const u8) usize;
extern fn isspace(ch: c_int) c_int;

// Zig std.c imports
extern fn fabs(x: f64) f64;

// stdout
extern const __stdoutp: *anyopaque;

// Paint mode query from terminal_io.zig — when paint mode is active,
// we skip per-call fflush to allow output batching (BEGINPAINT/ENDPAINT).
extern fn basic_is_paint_mode() i32;

fn flushIfNeeded() void {
    if (basic_is_paint_mode() == 0) {
        _ = fflush(__stdoutp);
    }
}

// =========================================================================
// StringDescriptor — opaque to this module, accessed via string_to_utf8
// =========================================================================

const StringDescriptor = anyopaque;

// =========================================================================
// Internal helpers
// =========================================================================

/// Reverse a byte slice in place
fn reverseSlice(s: []u8) void {
    if (s.len <= 1) return;
    var i: usize = 0;
    var j: usize = s.len - 1;
    while (i < j) {
        const tmp = s[i];
        s[i] = s[j];
        s[j] = tmp;
        i += 1;
        j -= 1;
    }
}

/// Find length of a null-terminated C string in a buffer
fn cstrLen(buf: [*]const u8, max: usize) usize {
    var i: usize = 0;
    while (i < max and buf[i] != 0) : (i += 1) {}
    return i;
}

/// Check if a character is whitespace
fn isSpaceChar(ch: u8) bool {
    return ch == ' ' or ch == '\t' or ch == '\n' or ch == '\r';
}

/// Format a numeric value according to a mask pattern
fn formatNumeric(output: [*]u8, output_size: usize, mask: [*:0]const u8, value_str: [*:0]const u8) void {
    // Try to parse value as a number
    var endptr: [*:0]const u8 = undefined;
    const value = strtod(value_str, &endptr);
    const is_numeric = (endptr[0] == 0 or isspace(@intCast(endptr[0])) != 0);

    if (!is_numeric) {
        // Not a number - just copy the value
        _ = snprintf(output, output_size, "%s", value_str);
        return;
    }

    const is_neg = (value < 0);
    const abs_val = fabs(value);

    // Analyze mask features
    const mask_len = strlen(mask);
    const has_comma = (strchr(mask, ',') != null);
    const has_plus = (mask[0] == '+');
    const has_minus_suffix = (mask_len > 0 and mask[mask_len - 1] == '-');
    const has_exp = (strstr(mask, "^^^^") != null);
    const has_dollar = (strstr(mask, "$$") != null);
    const has_asterisk = (strstr(mask, "**") != null);

    // Determine precision
    var precision: c_int = 0;
    const dot = strchr(mask, '.');
    if (dot) |dot_ptr| {
        var p: [*:0]const u8 = @ptrCast(dot_ptr);
        p += 1; // skip '.'
        while (p[0] == '#' or p[0] == '^') {
            precision += 1;
            p += 1;
        }
    }

    // Core conversion
    var work: [128]u8 = undefined;
    if (has_exp) {
        _ = snprintf(&work, work.len, "%.*E", precision, value);
    } else {
        _ = snprintf(&work, work.len, "%.*f", precision, abs_val);
    }
    const work_len = cstrLen(&work, work.len);

    // Manual comma insertion (integer part only)
    var comma_buf: [128]u8 = undefined;
    if (has_comma and !has_exp) {
        // Find the dot in work
        var dot_pos: usize = work_len;
        for (work[0..work_len], 0..) |ch, idx| {
            if (ch == '.') {
                dot_pos = idx;
                break;
            }
        }
        const int_len = dot_pos;

        // Build reversed with commas
        var temp: [128]u8 = .{0} ** 128;
        var dst: usize = 0;
        var count: usize = 0;
        var src_i: usize = int_len;
        while (src_i > 0) {
            src_i -= 1;
            if (count > 0 and count % 3 == 0 and dst < temp.len - 1) {
                temp[dst] = ',';
                dst += 1;
            }
            if (dst < temp.len - 1) {
                temp[dst] = work[src_i];
                dst += 1;
            }
            count += 1;
        }
        temp[dst] = 0;
        reverseSlice(temp[0..dst]);

        // Append decimal part
        if (dot_pos < work_len) {
            var k: usize = dot_pos;
            while (k < work_len and dst < temp.len - 1) {
                temp[dst] = work[k];
                dst += 1;
                k += 1;
            }
            temp[dst] = 0;
        }

        @memcpy(comma_buf[0..dst], temp[0..dst]);
        comma_buf[dst] = 0;
        @memcpy(work[0 .. dst + 1], comma_buf[0 .. dst + 1]);
    }

    // Decorations (Sign, $, *)
    var decorated: [256]u8 = .{0} ** 256;
    var dec_pos: usize = 0;

    // Prefix: sign
    if (has_plus) {
        if (is_neg) {
            decorated[dec_pos] = '-';
        } else {
            decorated[dec_pos] = '+';
        }
        dec_pos += 1;
    } else if (is_neg and !has_minus_suffix) {
        decorated[dec_pos] = '-';
        dec_pos += 1;
    }

    // Prefix: dollar
    if (has_dollar) {
        decorated[dec_pos] = '$';
        dec_pos += 1;
    }

    // Copy work string
    const final_work_len = cstrLen(&work, work.len);
    for (work[0..final_work_len]) |ch| {
        if (dec_pos < decorated.len - 1) {
            decorated[dec_pos] = ch;
            dec_pos += 1;
        }
    }

    // Suffix: trailing minus
    if (is_neg and has_minus_suffix) {
        if (dec_pos < decorated.len - 1) {
            decorated[dec_pos] = '-';
            dec_pos += 1;
        }
    }
    decorated[dec_pos] = 0;

    // Padding logic
    const actual_len = dec_pos;
    if (actual_len > mask_len) {
        // Overflow - prefix with %
        output[0] = '%';
        var i: usize = 0;
        while (i < actual_len and i + 1 < output_size - 1) {
            output[i + 1] = decorated[i];
            i += 1;
        }
        output[i + 1] = 0;
    } else {
        const pad = mask_len - actual_len;
        const pad_char: u8 = if (has_asterisk) '*' else ' ';
        var i: usize = 0;
        while (i < pad and i < output_size - 1) {
            output[i] = pad_char;
            i += 1;
        }
        var j: usize = 0;
        while (j < actual_len and i < output_size - 1) {
            output[i] = decorated[j];
            i += 1;
            j += 1;
        }
        output[i] = 0;
    }
}

/// Extract a format pattern starting at position p.
/// Returns the length of the pattern (0 if not a pattern).
fn extractPattern(p: [*:0]const u8, pattern: [*]u8, pattern_size: usize) usize {
    var offset: usize = 0;
    var len: usize = 0;

    // Check for leading +, $$, or **
    if (p[0] == '+') {
        pattern[0] = '+';
        len = 1;
        offset = 1;
    } else if (p[0] == '$' and p[1] == '$') {
        pattern[0] = '$';
        pattern[1] = '$';
        len = 2;
        offset = 2;
    } else if (p[0] == '*' and p[1] == '*') {
        pattern[0] = '*';
        pattern[1] = '*';
        len = 2;
        offset = 2;
    }

    // Collect #, comma, and decimal point
    while (true) {
        const ch = p[offset];
        if (ch != '#' and ch != ',' and ch != '.') break;
        if (len < pattern_size - 1) {
            pattern[len] = ch;
            len += 1;
        }
        offset += 1;
    }

    // Check for ^^^^
    if (strncmp(@ptrCast(p + offset), "^^^^", 4) == 0) {
        if (len + 4 < pattern_size) {
            pattern[len] = '^';
            pattern[len + 1] = '^';
            pattern[len + 2] = '^';
            pattern[len + 3] = '^';
            len += 4;
        }
        offset += 4;
    }

    // Check for trailing -
    if (p[offset] == '-') {
        if (len < pattern_size - 1) {
            pattern[len] = '-';
            len += 1;
        }
        offset += 1;
    }

    pattern[len] = 0;

    // Return length only if we found at least one #
    var has_hash = false;
    for (pattern[0..len]) |ch| {
        if (ch == '#') {
            has_hash = true;
            break;
        }
    }
    // Also check for @ in original (though @ is handled separately)
    if (has_hash) return offset;

    return 0;
}

// =========================================================================
// Exported function
// =========================================================================

/// PRINT USING implementation with full numeric formatting support
export fn basic_print_using(format: ?*StringDescriptor, count: i64, args: ?[*]?*StringDescriptor) callconv(.c) void {
    const fmt_desc = format orelse return;

    // Extract format string UTF-8
    const fmt = string_to_utf8(fmt_desc);
    const fmt_copy = strdup(fmt) orelse {
        basic_error_msg("Out of memory in basic_print_using");
        return;
    };
    defer c.free(fmt_copy);

    // Collect argument UTF-8 strings
    const arg_count: usize = if (count > 0) @intCast(count) else 0;
    var arg_strings_buf: [64]?[*:0]u8 = .{null} ** 64;
    const effective_count = @min(arg_count, arg_strings_buf.len);

    if (effective_count > 0) {
        if (args) |arg_ptrs| {
            for (0..effective_count) |i| {
                if (arg_ptrs[i]) |arg| {
                    const s = string_to_utf8(arg);
                    arg_strings_buf[i] = strdup(s);
                }
            }
        }
    }
    defer {
        for (0..effective_count) |i| {
            if (arg_strings_buf[i]) |s| c.free(s);
        }
    }

    // Process the format string with collected arguments
    var argIndex: usize = 0;
    var p: [*:0]const u8 = fmt_copy;

    while (p[0] != 0) {
        if (p[0] == '@') {
            // String substitution
            if (argIndex < effective_count) {
                if (arg_strings_buf[argIndex]) |s| {
                    _ = printf("%s", s);
                }
            }
            argIndex += 1;
            p += 1;
        } else if (p[0] == '#' or p[0] == '+' or
            (p[0] == '$' and p[1] == '$') or
            (p[0] == '*' and p[1] == '*'))
        {
            var pattern: [128]u8 = undefined;
            const pattern_len = extractPattern(p, &pattern, pattern.len);

            if (pattern_len > 0) {
                if (argIndex < effective_count) {
                    if (arg_strings_buf[argIndex]) |arg_str| {
                        var formatted: [256]u8 = undefined;
                        formatNumeric(&formatted, formatted.len, @ptrCast(&pattern), arg_str);
                        _ = printf("%s", &formatted);
                    }
                }
                argIndex += 1;
                p += pattern_len;
            } else {
                _ = putchar(@intCast(p[0]));
                p += 1;
            }
        } else {
            _ = putchar(@intCast(p[0]));
            p += 1;
        }
    }

    flushIfNeeded();
}
